pkg/operator/staticpod/controller/installer/installer_controller.go

```diff
@@ -674,40 +742,78 @@ func (c *InstallerController) newNodeStateForInstallInProgress(ctx context.Conte
-				errors = append(errors, fmt.Sprintf("no detailed termination message, see `oc get -oyaml -n %q pods %q`", ns, name))
+			if len(staticPodErrors) == 0 {
+				staticPodErrors = append(staticPodErrors, fmt.Sprintf("no detailed termination message, see `oc get -oyaml -n %q pods %q`", ns, name))
+			}
+			ret.LastFailedRevisionErrors = staticPodErrors
+
+			if startupMonitor {
+				// this will be overridden again when the fallback pod is observed
+				ret.LastFailedRevisionErrors = append(ret.LastFailedRevisionErrors, "falling back to last-known-good revision")
+			} else {
+				ret.TargetRevision = 0 // stop installer retries, but keep running in startup-monitor case where we leave this to the startup-monitor to do
+				ret.LastFailedRevision = ret.TargetRevision
+
+				// increase failed count, but not in startup-monitor case where this is done when the fallback pod is observed below in staticPodStateFallback,
+				ret.LastFailedRevision = currNodeState.TargetRevision
+				if ret.LastFailedTime == nil || !ret.LastFailedTime.Time.Equal(ts) {
+					// avoid increasing the count continuously by comparing the fail time
+					failTime := metav1.NewTime(ts)
+					ret.LastFailedTime = &failTime
+					ret.LastFailedCount++
+				}
 			}
-			ret.LastFailedRevisionErrors = errors
 
-			return ret, false, fmt.Sprintf("operand pod failed: %v", strings.Join(errors, "\n")), nil
+			return ret, false, fmt.Sprintf("operand pod failed: %v", strings.Join(staticPodErrors, "\n")), nil
 
 		case staticPodStateReady:
 			if currNodeState.TargetRevision > ret.CurrentRevision {
 				ret.CurrentRevision = currNodeState.TargetRevision
 			}
-			ret.TargetRevision = 0
-			ret.LastFailedRevision = 0
-			ret.LastFailedTime = nil
-			ret.LastFailedCount = 0
-			ret.LastFailedRevisionErrors = nil
+
+			// in case of startup-monitor we leave marking a ready operand as current version to the startup-monitor.
+			//
+			// Two edge case:
+			// 1. startup-monitor is enabled. A new operand is rolled out with the startup-monitor. It will mark the new revision as current.
+			// 2. startup-monitor is disabled. A new operand is rolled out without the startup-monitor pod. The operator will now mark the revision as current.
+			// 3. startup-monitor is disabled, while a new operator WITH startup-monitor is rolled out. The upper revision override will mark the revision as current.
+			if !startupMonitor {
+				ret.TargetRevision = 0
+				ret.LastFailedRevision = 0
+				ret.LastFailedTime = nil
+				ret.LastFailedCount = 0
+				ret.LastFailedRevisionErrors = nil
+			}
 			return ret, false, staticPodReason, nil
+
+		case staticPodStateFallback:
+			reason = staticPodReason
+			if ret.LastFailedTime == nil || !ret.LastFailedTime.Time.Equal(ts) { // do this only once per fallback
+				failTime := metav1.NewTime(ts)
+				ret.LastFailedTime = &failTime
+				ret.LastFailedRevision = ret.TargetRevision
+				ret.LastFailedRevisionErrors = staticPodErrors
+				ret.LastFailedCount++
+			}
+
```

Comment:
no `return` statement?

Reply:
I got rid of the return at the very end of the func and added direct returns everywhere else. Much clearer IMO.